design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,461 @@
+           
+## Motivation and background
+
+Data stored in Kafka has a key and a value. These can be serialized using different formats, but are
+generally use a common format. ksqlDB supports multiple _value_ [data formats][1], but requires the 
+key data format to be the `KAFKA`.  
+
+This limitation is particularly problematic for tables. ksqlDB is unable to access changelog topics
+in Kafka that have non-`KAFKA` formatted keys. As the key of the Kafka record is the `PRIMARY KEY` 
+of the table, it is essential that the key can be read if the changelog is to be materialised into a 
+table. When changelog topics have non-`KAFKA` key formats, the limitation precludes ksqlDB as a solution.
+
+The limitation is less damaging for streams. However, it is still the case that the user looses the
+ability to access the data in the Kafka record's key. If this data is not duplicated in the record's
+value, which generally seems to be the case, then the data is not accessible at all. If the data is
+required, then the limitation precludes ksqlDB as a solution.
+
+As well as unsupported _input_ key formats, ksqlDB is equally precluded should a solution require the 
+_output_ to have a non-`KAFKA` key format. ksqlDB is already often used as the glue between 
+disparate systems, even though it is limited to changing the _values_ format and structure.
+Supporting other key formats opens this up to also transforming the key into a different format. 
+
+In some cases users are able to work around this limitation. This may involve changing upstream code,
+or introducing pre-processing, or, in the case of Connect, using SMTs to convert the key format. All
+such solutions tend to increase the complexity of the system, and generally hurt performance.
+
+To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 
+
+## What is in scope
+
+ * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+ * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+ * Addition of new server configuration to provide defaults for key format.
+ * Support for additional key column data types, where the key format supports it:
+    * `DECIMAL`
+    * `BOOLEAN`
+ * Support of the following key formats:
+    * `KAFKA`: the current key format.
+    * `DELIMITED`: single key columns as a single string value.
+    * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.
+    * `AVRO`: single key column as an anonymous value, i.e. not within an Avro record.
+    * `NULL`: special format indicating no data, or ignored data, e.g. a key-less stream.
+  * Full support of these key formats for all supported SQL syntax.
+  * Automatic repartitioning of streams and tables for joins where key formats do not match.
+  * Support for reading & writing key schemas to & from the schema registry. 
```

Comment:
todo: Enhancing QTT, testing too ... and new SQL testing tool?

Reply:
If I've been coding it as I intended, YATT should get this KLIP for free 😂